### PR TITLE
[4.0] [Workflow] Implement the whole permission system for the workflow

### PR DESCRIPTION
**Pull request by bembelimen**
### Summary of Changes
In the first step, the permissions of changing the workflow were very soft, so with this PR the permission checks are harden.

See:

- https://github.com/joomla/joomla-cms/pull/29235#issuecomment-635245470
- https://github.com/bembelimen/joomla-cms/issues/59

### Testing Instructions
Create 2 accounts:
- a full Super User account to change permissions (account 1)
- a "normal" backend account to test (account 2)

Now set up the com_content permissions with account 1, that account 2 can access workflow or not (+ test). Then activate/deactivate all permissions one by one (create, edit, delete) and test the workflow elements, if the set permission takes action.

After that go into the individual elements and test the permissions there, too.

### CODE REVIEW

File: administrator/components/com_workflow/src/Table/WorkflowTable.php [hunk 1]
```diff
@@ -156,10 +157,14 @@ public function check()
 
 			$state = $db->setQuery($query)->loadObject();
 
-			if (empty($state) || $state->id === $this->id)
+			// We have nowhere a default workflow => set current to default to recover
```
Comment (brianteeman) [inline administrator/components/com_workflow/src/Table/WorkflowTable.php]:
```suggestion
			// If there is no default workflow => set the current to default to recover
```

File: administrator/components/com_workflow/src/Table/WorkflowTable.php [hunk 2]
```diff
@@ -156,10 +157,14 @@ public function check()
...
+			if (empty($state))
 			{
 				$this->default = '1';
-
+			}
+			// This workflow was default, but someone try to disable it => not allowed
```
Comment (brianteeman) [inline administrator/components/com_workflow/src/Table/WorkflowTable.php]:
```suggestion
			// This workflow is the default, but someone has tried to disable it => not allowed
```

**Comment by bembelimen** [comment]
Thank you!

**Comment by HLeithner** [comment]
@bembelimen please fix cs https://ci.joomla.org/joomla/joomla-cms/32784/1/9

**Comment by bembelimen** [comment]
It's already fixed locally, just don't want to push every minute some updates, so I collect and push every few hours.

**Comment by Quy** [comment]
Edit a transition.

> Undefined index: workflow_id in \administrator\components\com_workflow\src\Model\TransitionModel.php on line 262

**Comment by bembelimen** [comment]
> 
> 
> Edit a transition.
> 
> > Undefined index: workflow_id in \administrator\components\com_workflow\src\Model\TransitionModel.php on line 262

Thx

**Comment by Quy** [comment]
Edit and save a transition.

> PHP Notice:  Undefined variable: pk in \administrator\components\com_workflow\src\Model\TransitionModel.php on line 143

**Review by wilsonge — APPROVED** [review]

File: administrator/components/com_workflow/src/Table/StageTable.php [hunk 3]
```diff
@@ -192,6 +187,31 @@ public function store($updateNulls = true)
 		return parent::store($updateNulls);
 	}
 
+	/**
+	 * Method to bind an associative array or object to the Table instance.This
+	 * method only binds properties that are publicly accessible and optionally
```
Comment (Quy) [inline administrator/components/com_workflow/src/Table/StageTable.php]:
```suggestion
	 * Method to bind an associative array or object to the Table instance.
	 * This method only binds properties that are publicly accessible and optionally
```

File: administrator/components/com_workflow/src/Table/TransitionTable.php [hunk 4]
```diff
@@ -50,8 +51,31 @@ class TransitionTable extends Table
 	public function __construct(DatabaseDriver $db)
 	{
 		parent::__construct('#__workflow_transitions', 'id', $db);
+	}
 
-		$this->access = (int) Factory::getApplication()->get('access');
+	/**
+	 * Method to bind an associative array or object to the Table instance.This
+	 * method only binds properties that are publicly accessible and optionally
```
Comment (Quy) [inline administrator/components/com_workflow/src/Table/TransitionTable.php]:
```suggestion
	 * Method to bind an associative array or object to the Table instance.
	 * This method only binds properties that are publicly accessible and optionally
```

File: administrator/components/com_workflow/src/Table/WorkflowTable.php [hunk 5]
```diff
@@ -230,6 +228,31 @@ public function store($updateNulls = true)
 		return parent::store($updateNulls);
 	}
 
+	/**
+	 * Method to bind an associative array or object to the Table instance.This
+	 * method only binds properties that are publicly accessible and optionally
```
Comment (Quy) [inline administrator/components/com_workflow/src/Table/WorkflowTable.php]:
```suggestion
	 * Method to bind an associative array or object to the Table instance.
	 * This method only binds properties that are publicly accessible and optionally
```

**Comment by bembelimen** [comment]
Thx @Quy

**Comment by softforge** [comment]
I have tested this item :white_check_mark: successfully on 6161ff0ef81ab81be28a619b5a66df8395478fa8<br /><br />Thanks to Benjamin for helping me through this one. We discovered an interesting quirk with the display but the permissions worked both in the overall and in the specifics for each workflow<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29263">issues.joomla.org/tracker/joomla-cms/29263</a>.</sub>

**Comment by particthistle** [comment]
I have tested this item :white_check_mark: successfully on 6161ff0ef81ab81be28a619b5a66df8395478fa8<br /><br />Tested several permission variations

- Viewing Workflows
- Creating Workflows
- Deleting Workflows<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29263">issues.joomla.org/tracker/joomla-cms/29263</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29263">issues.joomla.org/tracker/joomla-cms/29263</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**